src/NHibernate/Engine/TableGroupJoinHelper.cs

```diff
@@ -0,0 +1,134 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using NHibernate.Persister.Entity;
+using NHibernate.SqlCommand;
+
+namespace NHibernate.Engine
+{
+	//Generates table group join if neccessary. Example of generated query with table group join:
+	// SELECT *
+	// FROM Person person0_
+	// INNER JOIN (
+	// IndividualCustomer individual1_ 
+	// INNER JOIN Customer individual1_1_ ON individual1_.IndividualCustomerID = individual1_1_.Id
+	// ) ON person0_.Id = individual1_.PersonID AND individual1_1_.Deleted = @p0
+	internal class TableGroupJoinHelper
+	{
+		internal static bool ProcessAsTableGroupJoin(IReadOnlyList<IJoin> tableGroupJoinables, SqlString[] withClauseFragments, bool includeAllSubclassJoins, JoinFragment joinFragment, Func<string, bool> isSubclassIncluded, ISessionFactoryImplementor sessionFactoryImplementor)
+		{
+			if (!TableGroupJoinHelper.NeedsTableGroupJoin(tableGroupJoinables, withClauseFragments, includeAllSubclassJoins))
+				return false;
+
+			var first = tableGroupJoinables[0];
+			string joinString = ANSIJoinFragment.GetJoinString(first.JoinType);
+			joinFragment.AddFromFragmentString(
+				new SqlString(
+					joinString,
+					" (",
+					first.Joinable.TableName,
+					" ",
+					first.Alias
+				));
+
+			foreach (var join in tableGroupJoinables)
+			{
+				if (join != first)
+					joinFragment.AddJoin(
+						join.Joinable.TableName,
+						join.Alias,
+						join.LHSColumns,
+						JoinHelper.GetRHSColumnNames(join.AssociationType, sessionFactoryImplementor),
+						join.JoinType,
+						SqlString.Empty);
+
+				bool include = includeAllSubclassJoins && isSubclassIncluded(join.Alias);
+				// TODO (from hibernate): Think about if this could be made always true 
+				// NH Specific: made always true (original check: join.JoinType == JoinType.InnerJoin)
+				var innerJoin = true;
+				joinFragment.AddJoins(
+					join.Joinable.FromJoinFragment(join.Alias, innerJoin, include),
+					join.Joinable.WhereJoinFragment(join.Alias, innerJoin, include));
+			}
+
+			var withClause = TableGroupJoinHelper.GetTableGroupJoinWithClause(withClauseFragments, first, sessionFactoryImplementor);
+			joinFragment.AddFromFragmentString(withClause);
+			return true;
+		}
+
+		private static bool NeedsTableGroupJoin(IReadOnlyList<IJoin> joins, SqlString[] withClauseFragments, bool includeSubclasses)
+		{
+			// If the rewrite is disabled or we don't have a with clause, we don't need a table group join
+			if ( /*!collectionJoinSubquery ||*/ withClauseFragments.All(x => SqlStringHelper.IsEmpty(x)))
```

Comment:
Should we keep that dead code and comment about it? I understand it is very likely coming from Hibernate code, but I do not see any value in keeping it, especially since we are now refactoring it out of the file where it was in Hibernate. If it has still some value, we should make it more explicit, like maybe telling "todo : handle disabling in case of collection join", if it is something that is needed.

```suggestion
			// If we don't have a with clause, we don't need a table group join
			if (withClauseFragments.All(x => SqlStringHelper.IsEmpty(x)))
```

Reply:
Ok. Let's just get rid of it..